### ISSUE TITLE
Try out inference

### ISSUE THREAD

**Issue by aot29** [aot29/ecomon#39]
## 26.08.2025
Try out birdid_medium on 1205E
* gpuMode False, takes 10:25 min for a batch of 100
* found 19295 records

## 11.08.2025
Made merge request. Models run if image is built locally.

## 14.08.2025
* [x] Tag old version of model-zoo on server
* [x] Checkout merge and build new version
* [x] Restart Docker ensemble on server
* [ ] Refactor Dockerfiles for models #50, #51, #53
* [ ] Repair the way parameters are passed to model zoo from ecomon (in ecomon image) #52 
* [ ] Try out inference on 1205H, check logs for errors, and GPU usage 
* [ ] Check if results are shown in records tables

* `watch -n 0.5 nvidia-smi`
* `docker logs -f akwamo-webservice-next-worker-1`
* in .env, USE-CPU set to 1, 

**avesecho_v1.3.0**
* Command stderr: INFO: Created TensorFlow Lite XNNPACK delegate for CPU.
* 160 MiB GPU memory allocated, but 0% usage (as shown in nvidia-smi)
* pip show tflite-runtime-gpu (inside avesecho container) -> tflite runtime for GPU is not installed 
-> not using GPU, takes +- 3m20s per 100 files batch

**avesecho_v1.3.0_transformer**
* model_inference_site[1225]: Command stderr: No pretrained weights exist or were found for this model. Using random initialization. No pretrained weights exist or were found for this model. Using random initialization.
* INFO: Created TensorFlow Lite XNNPACK delegate for CPU.

**birdid-europe254-medium**
* Uses GPU, +- 2m30s per 100 files batch
* Uses 2570 MiB GPU menory
* The command generated is: 
```
docker run -v /var/run/docker.sock:/var/run/docker.sock --rm -v /mnt/local/ecomon/tmp/1221/inputPaths.txt:/app/inputPaths.txt -v /mnt/local/ecomon/tmp/1221:/output -v /mnt/akwamodata:/data ghcr.io/mfn-berlin/birdid-model-zoo:latest --gpuIx 0 --nCpuWorkers 16 --batchSize 32 -i /app/inputPaths.txt -m birdid-europe254-medium -o /output -ov /mnt/local/ecomon/tmp/1221 --removeTemporaryResultFile -chown 0:0 --f pkl --gpuIx 1 -on output
```
So the gpuIx  parameter is set 2x: one is coming from BirID-Model-Zoo/ModelConfig.json, the other is set in ecomon_next/src/backend/worker/model_inference_site_task.py. The one set in model_inference_site_task is read from ecomon_next/.env
So a decision is needed: set the gpu in .env or in ModelConfig.json?

**birdid-europe254-large**
* Uses GPU, +- 2m30 per 100 files batch
* Uses 1680 MiB GPU memory
* The command generated is
```
docker run -v /var/run/docker.sock:/var/run/docker.sock --rm -v /mnt/local/ecomon/tmp/1223/inputPaths.txt:/app/inputPaths.txt -v /mnt/local/ecomon/tmp/1223:/output -v /mnt/akwamodata:/data ghcr.io/mfn-berlin/birdid-model-zoo:latest -i /app/inputPaths.txt -m birdid-europe254-large -o /output -ov /mnt/local/ecomon/tmp/1223 --removeTemporaryResultFile -chown 0:0 --f pkl --gpuIx 1 -on output
```

**birdnet**
* WARNING/ForkPoolWorker-2] model_inference_site[1222]: Command stderr: 2025-08-18 05:44:53.929845: I external/local_tsl/tsl/cuda/cudart_stub.cc:31] Could not find cuda drivers on your machine, GPU will not be used.
* Not using GPU, nor GPU memory. takes +- 1m05 per batch of 100 files

## Issue
The goal is to produce 1 year of inference for 1 location, using 2 models. Check the data and view it as tables in the web interface data tables/records page.

see: https://github.com/aot29/ecomon/wiki/DB%20Schema

## Preparation

* [x] Cleanup db of any inferences related to test site
  * [x] Make a backup of the database
  * [x] Import the database locally
  * [x] Attempt to cleanup 
* [x] Run birdid medium on the test data, look in the log for errors and warnings
* [x] Run birdnet on the test data, look in the log for errors and warnings
* [x] Check that inference results from the test data is correct in the database
* [x] Check that the data can be seen in the UI

## Test run
Test using test site

Cleanup DB
```
docker exec -it akwamo-webservice-next-db-1 bash
psql -h localhost -p 5432 -U ecomon -d ecomon
delete from model_inference_logs where record_id in (select id from records where site_id=1);
delete from model_inference_results where record_id in (select id from records where site_id=1);
```
Go to UI, run inference on test site, one model at a time, check results with:
`docker logs -f akwamo-webservice-next-worker-1`

| Model                 |Results               | Time      | Completed |
| -----------------  | ------------------ | ---------| ------------ | 
|avesecho_v1.3.0  | Model runs, but fails with DB error| n/a | No |
| avesecho_v1.3.0_transformer | Model runs, but fails with DB error| n/a | No |
| birdid-europe254-large | Processing compreted, skipped 1 file | 7 m | Yes |
| birdid-europe254-medium |  Processing completed, skipped 1 file | 3 m | Yes |
| birdnet_v2.2 | Processed all files | < 1 m | Yes |
| birdnet_v2.4 | Processed all files | < 1m | Yes |

Remarks:
* All models run on CPU, not on GPU " Could not find cuda drivers on your machine, GPU will not be used.". The machine has 2 GPUs, but GPU-Util is alwas 0%. This is probably a miconfiguration of the Docker container.
* avesecho (1.3.0 and 1.3.0_transformer) runs, but fails with DB error
* birdid (medium, large) runs, but skips some files
* birdnet (2.2, 2.4) runs
* Running several inferences with the same model simultaneously fails with a container naming conflict

##  Choose a location, a year, and run the 2 models, check the data in the database
Try out with: AKWAMO1205H (Neusorgefeld, BBNSF_1205H), 19241 records

* [x] How long does each model take?
* [x] Which models produce mistakes?

Results:
| Model                 |Results               | # Records|
| -----------------  | ------------------| ------------ | 
| avesecho_v1.3.0 |  Fails with DB error  | 0 | 
| avesecho_v1.3.0_transformer |  Fails with DB error  | 0 | 
| birdid-europe254-medium |Done | 45983625 | 
| birdid-europe254-large | Done|22591421|
| birdnet_v2.2 | Done | 5868497 |
| birdnet_v2.4 | Done | 3946986 |

**Comment by aot29** [comment]
Frage für Mario: Birdnet produziet Ergebnisse mit niedrigere Konfidenzwerte, selbst bei lauten Signalen. dadurch werden im Endeffekt wenigere Ergebnisse in der "Records" Tabelle angezeigt, wenn man einen Threshold (z.B. 0,5) setzt. 
Ist es möglich, die Konfidenzwerte zu normalisieren, so dass Birdnet (aber auch andere Modelle), gleich wahrscheinlich in der Ergebnisstabell auftauchen? 
Sieh hier für en Beispiel: https://129.70.51.241/ecomon_next/records/526649 (sortiere nach Model ID)

**Comment by aot29** [comment]
Command: 
https://github.com/MfN-Berlin/BirdID-Model-Zoo/blob/main/runModel.py
python -c "from runModel import *; getModelResults('birdnet_v2.4', '~/Desktop/inference-results/', './Data/TEST/TEST_20230506/')"
(ohne GPU)

birdnet_v2.4 (63.43s)
docker run --name bmz_birdnet_v2.4 --rm -v ./Data/TEST/TEST_20230506/:/input -v ~/Desktop/inference-results/birdnet_v2.4/:/output --ipc=host --shm-size=4g ghcr.io/mfn-berlin/birdnet-v24 -m birdnet_analyzer.analyze --i input --o output --overlap 1.0 --rtype csv --min_conf 0.01 --threads 8 --batchsize 16

birdnet_v2.2 (69.01s)
docker run --name bmz_birdnet_v2.2 --rm -v ./Data/TEST/TEST_20230506/:/input -v ~/Desktop/inference-results/birdnet_v2.2/:/output --ipc=host --shm-size=4g ghcr.io/mfn-berlin/birdnet-v22 -m birdnet_analyzer.analyze --i input --o output --overlap 1.0 --rtype csv --min_conf 0.01 --threads 8 --batchsize 16

avesecho_v1.3.0 (30.37s)
docker run --name bmz_avesecho_v1.3.0 --rm -v ./Data/TEST/TEST_20230506/:/app/audio -v ~/Desktop/inference-results/avesecho_v1.3.0/:/app/outputs --ipc=host --shm-size=4g registry.gitlab.com/arise-biodiversity/dsi/algorithms/avesecho-v1/avesechov1:v1.3.0 --i audio --model_name 'fc' --add_csv --mconf 0.01

avesecho_v1.3.0_transformer (105.06s)
docker run --name bmz_avesecho_v1.3.0_transformer --rm -v ./Data/TEST/TEST_20230506/:/app/audio -v ~/Desktop/inference-results/avesecho_v1.3.0_transformer/:/app/outputs --ipc=host --shm-size=4g registry.gitlab.com/arise-biodiversity/dsi/algorithms/avesecho-v1/avesechov1:v1.3.0 --i audio --model_name 'passt' --add_csv --mconf 0.01

birdid-europe254-medium (75.85s)
docker run --name bmz_birdid-europe254-medium --rm -v ./Data/TEST/TEST_20230506/:/input -v ~/Desktop/inference-results/birdid-europe254-medium/:/output --ipc=host --shm-size=4g ghcr.io/mfn-berlin/birdid-europe254-v250331-1:311173148b6396f89a9d4b104cad95c064697fda python inference.py -i /input -o /output --fileOutputFormats labels_csv --overlapInPerc 60.0 --csvDelimiter , --sortSpecies --nameType sci --includeFilePathInOutputFiles --modelSize medium --debug --minConfidence 0.01 --batchSizeFiles 16 --nCpuWorkers 8 --batchSizeInference 16

birdid-europe254-large (?)
docker run --name bmz_birdid-europe254-large --rm -v ./Data/TEST/TEST_20230506/:/input -v ~/Desktop/inference-results/birdid-europe254-large/:/output --ipc=host --shm-size=4g ghcr.io/mfn-berlin/birdid-europe254-v250331-1:311173148b6396f89a9d4b104cad95c064697fda python inference.py -i /input -o /output --fileOutputFormats labels_csv --overlapInPerc 60.0 --csvDelimiter , --sortSpecies --nameType sci --includeFilePathInOutputFiles --modelSize large --minConfidence 0.01 --batchSizeFiles 16 --nCpuWorkers 8 --batchSizeInference 16
